### PR TITLE
Fix:  creating gsi  under on-demand mode

### PR DESCRIPTION
**Pull request by Julian-Chu**
Issue #, if available:

Description of changes:

- fix empty NonKeyAttributes validation issue
- fix GSI creation issue when using `pay-per-request` mode
```shell
2023-04-19T17:31:29.932+0200    ERROR   Reconciler error        {"controller": "table", "controllerGroup": "dynamodb.services.k8s.aws", "controllerKind": "Table", "Table": {"name":"ack-demo-table-provisioned-global-secondary-indexes","namespace":"ack-system"}, "namespace": "ack-system", "name": "ack-demo-table-provisioned-global-secondary-indexes", "reconcileID": "c7354a77-56c4-4c56-b715-f16b59d1926e", "error": "ValidationException: One or more parameter values were invalid: Neither ReadCapacityUnits nor WriteCapacityUnits can be specified when BillingMode is PAY_PER_REQUEST\n\tstatus code: 400, request id: R901I15N84RUVS24D4HVRP6L3JVV4KQNSO5AEMVJF66Q9ASUAAJG"}

2023-04-19T18:54:54.887+0200    ERROR   Reconciler error        {"controller": "table", "controllerGroup": "dynamodb.services.k8s.aws", "controllerKind": "Table", "Table": {"name":"ack-demo-table-provisioned-global-secondary-indexes","namespace":"ack-system"}, "namespace": "ack-system", "name": "ack-demo-table-provisioned-global-secondary-indexes", "reconcileID": "223f9f39-f2cc-460c-ad78-56fce55b4bf3", "error": "ValidationException: One or more parameter values were invalid: ProvisionedThroughput should not be specified for index: id-index when BillingMode is PAY_PER_REQUEST\n\tstatus code: 400, request id: JO5OGQR8IN74JJ2C62O43ESU43VV4KQNSO5AEMVJF66Q9ASUAAJG"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
```
```yaml
    conditions:
    - message: |
        InvalidParameter: 3 validation error(s) found.
        - minimum field size of 1, UpdateTableInput.GlobalSecondaryIndexUpdates[0].Create.Projection.NonKeyAttributes.
        - minimum field value of 1, UpdateTableInput.GlobalSecondaryIndexUpdates[0].Create.ProvisionedThroughput.ReadCapacityUnits.
        - minimum field value of 1, UpdateTableInput.GlobalSecondaryIndexUpdates[0].Create.ProvisionedThroughput.WriteCapacityUnits.
      status: "True"
      type: ACK.Terminal
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/resource/table/hooks_global_secondary_indexes.go [hunk 1]
```diff
@@ -233,21 +233,21 @@ func (rm *resourceManager) newUpdateTableGlobalSecondaryIndexUpdatesPayload(
...
-			provisionedThroughput.WriteCapacityUnits = aws.Int64(*pt.WriteCapacityUnits)
-		} else {
-			provisionedThroughput.WriteCapacityUnits = aws.Int64(0)
-		}
-	} else {
-		provisionedThroughput.ReadCapacityUnits = aws.Int64(0)
-		provisionedThroughput.WriteCapacityUnits = aws.Int64(0)
+	if pt == nil {
+		return nil
+	}
```
Comment (a-hilaly) [inline pkg/resource/table/hooks_global_secondary_indexes.go:238]: Nice catch sir!

File: pkg/resource/table/hooks_global_secondary_indexes.go [hunk 2]
```diff
@@ -233,21 +233,21 @@ func (rm *resourceManager) newUpdateTableGlobalSecondaryIndexUpdatesPayload(
...
+		// Minimum capacity units is 1 when using provisioned capacity mode
+		ReadCapacityUnits:  aws.Int64(1),
+		WriteCapacityUnits: aws.Int64(1),
+	}
+	if pt.ReadCapacityUnits != nil {
+		provisionedThroughput.ReadCapacityUnits = aws.Int64(*pt.ReadCapacityUnits)
+	}
+
+	if pt.WriteCapacityUnits != nil {
+		provisionedThroughput.WriteCapacityUnits = aws.Int64(*pt.WriteCapacityUnits)
```
Comment (a-hilaly) [inline pkg/resource/table/hooks_global_secondary_indexes.go:250]: Thoughts: What if we're not using PROVISIONED capacity mode? Would 0's be allowed? If that's the case would it be worth to evaluate the capacity mode in this function?

File: pkg/resource/table/hooks_global_secondary_indexes.go [hunk 3]
```diff
@@ -263,12 +263,10 @@ func newSDKProjection(p *v1alpha1.Projection) *svcsdk.Projection {
 		}
 		if p.NonKeyAttributes != nil {
 			projection.NonKeyAttributes = p.NonKeyAttributes
-		} else {
-			projection.NonKeyAttributes = []*string{}
 		}
 	} else {
 		projection.ProjectionType = aws.String("")
-		projection.NonKeyAttributes = []*string{}
+		projection.NonKeyAttributes = nil
```
Comment (a-hilaly) [inline pkg/resource/table/hooks_global_secondary_indexes.go:269]: Another nice catch!

**Review by a-hilaly — APPROVED** [review]
Thank you very much @Julian-Chu ! This is neat!
I have few questions below.

File: pkg/resource/table/hooks_global_secondary_indexes.go (hunk 2 above)
Comment (Julian-Chu) [reply pkg/resource/table/hooks_global_secondary_indexes.go:250]: I think you mean the `PAY_PER_REQUEST` mode. 
this is tricky part. IIRC,  we don't allow to set 0 capacity units for `PAY_PER_REQUEST` , please see the following message
```yaml
   conditions:
    - message: |
        InvalidParameter: 3 validation error(s) found.
        - minimum field size of 1, UpdateTableInput.GlobalSecondaryIndexUpdates[0].Create.Projection.NonKeyAttributes.
        - minimum field value of 1, UpdateTableInput.GlobalSecondaryIndexUpdates[0].Create.ProvisionedThroughput.ReadCapacityUnits.
        - minimum field value of 1, UpdateTableInput.GlobalSecondaryIndexUpdates[0].Create.ProvisionedThroughput.WriteCapacityUnits.
      status: "True"
      type: ACK.Terminal
```  

but aws DDB api will set capacity units as 0 when `PAY_PER_REQUEST`




> If you set BillingMode as PROVISIONED, you must specify this property. If you set BillingMode as PAY_PER_REQUEST, you cannot specify this property.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-dynamodb-table.html#cfn-dynamodb-table-provisionedthroughput

> If read/write capacity mode is PAY_PER_REQUEST the value is set to 0.

https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_ProvisionedThroughput.html

**Review by a-hilaly — APPROVED** [review]
👍 
/lgtm

**Pull request merged**